README
Flask-Dogpile-Cache
-------------------


Easy to Use
```````````

.. code:: python

    # ==================== config.py ====================

    DOGPILE_CACHE_URLS = ["127.0.0.1:11211"]
    DOGPILE_CACHE_REGIONS = [
        ('hour', 3600),
        ('day', 3600 * 24),
        ('month', 3600 * 24 * 31),
    ]
    # Default settings:
    # DOGPILE_CACHE_BACKEND = 'dogpile.cache.memcached'


    # ==================== app.py ====================

    import config
    from flask import Flask
    from flask.ext.dogpile_cache import DogpileCache

    app = Flask(__name__)
    app.config.from_object(config)

    cache = DogpileCache()
    cache.init_app(app)
    # Alternative way: cache = DogpileCache(app)

    @cache.region('hour')
    def cached_func(*args):
        print "Heavy computation here", args
        return args

    value = cached_func()

    cache.invalidate(cached_func, *args)  # Invalidating cache for cached_func
    cache.refresh(cached_func, *args)     # Refreshing cache for cached_func
    cache.set(cached_func, value, *args)  # Setting custom value for cached_func
    cache.invalidate_region('hour')       # Invalidate cache for all funcs
                                          # decorated with @cache.region('hour')
    cache.invalidate_all_regions()        # Invalidate cache for all funcs
                                          # decorated with @cache.region


Easy to Install
```````````````

.. code:: bash

    $ pip install Flask-Dogpile-Cache

Links
`````

* `development version
  <http://bitbucket.org/ponomar/flask-dogpile-cache>`_


CHANGES
Changelog
=========

Here you can see the full list of changes between each Flask-Dogpile-Cache
release.

Version 0.2
-----------
- Add tests.
- Add methods `cache.invalidate_region()`, `cache.invalidate_all_regions()`.
- Add __doc__ to DogpileCache class.

Version 0.1
-----------
- Initial public release


tests.py
from __future__ import with_statement

import sys

from copy import deepcopy
from flask import Flask
from flask.ext.dogpile_cache import DogpileCache

if sys.version_info < (2, 7):
    import unittest2 as unittest
else:
    import unittest


class DogpileCacheTest(unittest.TestCase):
    def setUp(self):
        default_cache_backend = 'dogpile.cache.memcached'
        default_cache_urls = ["127.0.0.1:11211"]
        self.config = dict(
            DOGPILE_CACHE_BACKEND=default_cache_backend,
            DOGPILE_CACHE_URLS=default_cache_urls,
            DOGPILE_CACHE_REGIONS=[
                ('hour', 3600, default_cache_backend, default_cache_urls),
                ('day', 3600 * 24),
                ('week', 3600 * 24 * 7),
                ('month', 3600 * 24 * 31),
            ],
            DOGPILE_CACHE_ARGUMENTS={'binary': True},
        )

        self.app = Flask(__name__)
        self.app.debug = True
        self.cache = DogpileCache(self.app, self.config)

        self.func_cached_for_hour_value = 12345
        self.func_cached_for_day_value = 67890

        @self.cache.region('hour')
        def func_cached_for_hour(delta=0):
            return self.func_cached_for_hour_value + delta

        @self.cache.region('day')
        def func_cached_for_day(delta=0):
            return self.func_cached_for_day_value + delta

        self.func_cached_for_hour = func_cached_for_hour
        self.func_cached_for_day = func_cached_for_day
        self.func_result_map = ((self.func_cached_for_hour,
                                 self.func_cached_for_hour_value),
                                (self.func_cached_for_day,
                                 self.func_cached_for_day_value))

    def tearDown(self):
        self.config = None
        self.app = None
        self.cache = None
        self.func_result_map = None
        self.func_cached_for_hour_value = None
        self.func_cached_for_day_value = None
        self.func_cached_for_hour = None
        self.func_cached_for_day = None

    def clean_up_cache(self):
        self.cache.invalidate_all_regions()

    def test_get(self):
        self.clean_up_cache()
        for func, correct_value in self.func_result_map:
            self.assertEqual(func(), correct_value)

    def test_get_with_arg(self):
        self.clean_up_cache()
        for func, correct_value in self.func_result_map:
            self.assertNotEqual(func(777), correct_value)

    def test_set(self):
        self.clean_up_cache()
        for func, correct_value in self.func_result_map:
            custom_value = correct_value + 1
            self.cache.set(func, custom_value)
            self.assertEqual(func(), custom_value)

    def test_refresh(self):
        self.clean_up_cache()
        for func, correct_value in self.func_result_map:
            custom_value = correct_value + 1
            self.cache.set(func, custom_value)
            self.assertEqual(func(), custom_value)
            self.cache.refresh(func)
            self.assertEqual(func(), custom_value)

    def test_invalidate(self):
        self.clean_up_cache()
        for func, correct_value in self.func_result_map:
            custom_value = correct_value + 1
            self.cache.set(func, custom_value)
            self.assertEqual(func(), custom_value)
            self.cache.invalidate(func)
            self.assertEqual(func(), correct_value)

    def test_invalidate_region(self):
        self.clean_up_cache()
        for func, correct_value in self.func_result_map:
            custom_value = correct_value + 1
            self.cache.set(func, custom_value)

        self.cache.invalidate_region('hour')
        self.assertEqual(self.func_cached_for_hour(),
                         self.func_cached_for_hour_value)
        self.assertNotEqual(self.func_cached_for_day(),
                            self.func_cached_for_day_value)

    def test_invalidate_all_regions(self):
        self.clean_up_cache()
        for func, correct_value in self.func_result_map:
            custom_value = correct_value + 1
            self.cache.set(func, custom_value)

        self.cache.invalidate_all_regions()
        for func, correct_value in self.func_result_map:
            self.assertEqual(func(), correct_value)

    def test_working_outside_of_application_context(self):
        cache = DogpileCache()
        self.assertRaises(RuntimeError, cache.get_region, 'x')
        self.assertRaises(RuntimeError, cache.get_all_regions)
        self.assertRaises(RuntimeError, cache.get_region_decorator, 'x')

    def test_cache_with_not_existent_region(self):
        @self.cache.region('not_existent_region_name')
        def func():
            pass

        self.assertRaises(KeyError, func)

    def test_not_cached_func(self):
        not_cached_func = lambda: None
        self.assertRaises(AttributeError,
                          self.cache.invalidate,
                          not_cached_func)
        self.assertRaises(AttributeError,
                          self.cache.refresh,
                          not_cached_func)
        self.assertRaises(AttributeError,
                          self.cache.set,
                          not_cached_func,
                          'some_value')

    def test_wrong_config(self):
        app = Flask(__name__)

        config = deepcopy(self.config)
        config['DOGPILE_CACHE_REGIONS'] = []
        self.assertRaises(ValueError, DogpileCache, app, config)

        config = deepcopy(self.config)
        config['DOGPILE_CACHE_REGIONS'].append(tuple())
        self.assertRaises(ValueError, DogpileCache, app, config)

        config = deepcopy(self.config)
        config['DOGPILE_CACHE_REGIONS'].append(('bad',))
        self.assertRaises(ValueError, DogpileCache, app, config)

        config = deepcopy(self.config)
        config['DOGPILE_CACHE_BACKEND'] = None
        self.assertRaises(ValueError, DogpileCache, app, config)

        config = deepcopy(self.config)
        config['DOGPILE_CACHE_URLS'] = None
        self.assertRaises(ValueError, DogpileCache, app, config)

        config = deepcopy(self.config)
        config['DOGPILE_CACHE_REGIONS'] = None
        self.assertRaises(ValueError, DogpileCache, app, config)

        config = deepcopy(self.config)
        config['DOGPILE_CACHE_ARGUMENTS'] = None
        self.assertRaises(ValueError, DogpileCache, app, config)


if __name__ == '__main__':
    unittest.main()


flask_dogpile_cache.py
from dogpile.cache import make_region
from functools import wraps
from types import NoneType


__version__ = '0.2'


class NotInitialized(object):
    pass


class DogpileCache(object):
    FUNC_REGION_NAME_ATTR = 'dogpile_cache_region_name'

    def __init__(self, app=None, config=None, wrappers_debug=None,
                 wrappers_production=None):
        """
        :param app: Flask app object.

        :param config: Configuration dict.

            DOGPILE_CACHE_BACKEND
                Optional. 'dogpile.cache.memcached' by default.
                You can override it with other available backends from
                dogpile.cache.

            DOGPILE_CACHE_URLS
                Required if not declared for each region.
                A list or tuple of cache servers' urls.
                For example, ["127.0.0.1:11211"].

            DOGPILE_CACHE_REGIONS
                Required. A list or tuple of region_data_tuples. Each
                region_data_tuple has from 2 to 4 elements:
                1) region_name (str, required). For example, 'hour'.
                2) region_timeout (int, required). For example, 3600.
                3) region_backend (str, optional). If not declared takes
                   its value from DOGPILE_CACHE_BACKEND.
                4) region_urls (list or tuple, optional). If not declared
                   takes its value from DOGPILE_CACHE_URLS.
                5) region_arguments (dict, optional). See description bellow.

            DOGPILE_CACHE_ARGUMENTS
                Optional.  The structure here is passed directly to the
                constructor of the class `CacheBackend` in use, though is
                typically a dictionary.

            The simplest config is:
                DOGPILE_CACHE_URLS = ["127.0.0.1:11211"]
                DOGPILE_CACHE_REGIONS = [
                    ('hour', 3600),
                    ('day', 3600 * 24),
                ]

        :param wrappers_debug, wrappers_production:
            A decorator class for altering the functionality of backends.
            Example:

                from dogpile.cache.proxy import ProxyBackend

                class DebugProxy(ProxyBackend):
                    @timer  # custom decorator that logs calls duration
                    def get(self, key):
                        return self.proxied.get(key)

                    @timer
                    def set(self, key, value):
                        return self.proxied.set(key, value)

                    @timer
                    def delete(self, key):
                        return self.proxied.delete(key)

        """
        self.config = config
        self.app = app
        self._wrappers_debug = wrappers_debug
        self._wrappers_production = wrappers_production
        self._cache_regions = NotInitialized()
        self._cache_regions_decorators = NotInitialized()

        if app is not None:
            self.init_app(app, config)

    def init_app(self, app, config=None, wrappers_debug=None,
                 wrappers_production=None):
        if not isinstance(config, (NoneType, dict)):
            raise ValueError("`config` must be dict or NoneType")

        if config is None:
            config = self.config
            if config is None:
                config = app.config

        config.setdefault('DOGPILE_CACHE_BACKEND', 'dogpile.cache.memcached')
        config.setdefault('DOGPILE_CACHE_URLS', None)
        config.setdefault('DOGPILE_CACHE_ARGUMENTS', {})
        if not (
            isinstance(config['DOGPILE_CACHE_REGIONS'], (list, tuple))
            and config['DOGPILE_CACHE_REGIONS']
        ):
            raise ValueError('`DOGPILE_CACHE_REGIONS` must be list or tuple')
        if not isinstance(config['DOGPILE_CACHE_ARGUMENTS'], dict):
            raise ValueError('`DOGPILE_CACHE_ARGUMENTS` must be dict')

        wrappers_debug = wrappers_debug or self._wrappers_debug
        wrappers_production = wrappers_production or self._wrappers_production

        self._set_cache_regions(app=app,
                                config=config,
                                wrappers_debug=wrappers_debug,
                                wrappers_production=wrappers_production)

    def _set_cache_regions(self, app, config, wrappers_debug,
                           wrappers_production):
        self._cache_regions = dict()
        self._cache_regions_decorators = dict()

        for region_tuple in config['DOGPILE_CACHE_REGIONS']:
            if len(region_tuple) < 2:
                raise ValueError('`DOGPILE_CACHE_REGIONS` tuple item length '
                                 'must be at least 2: region_name and '
                                 'region_timeout')

            region_name = region_tuple[0]
            region_timeout = region_tuple[1]

            if len(region_tuple) > 2:
                region_backend = region_tuple[2]
            elif config['DOGPILE_CACHE_BACKEND']:
                region_backend = config['DOGPILE_CACHE_BACKEND']
            else:
                raise ValueError('`DOGPILE_CACHE_BACKEND` must be specified '
                                 'or not initialized at all '
                                 '(for default value)')

            if len(region_tuple) > 3:
                region_urls = region_tuple[3]
            elif config['DOGPILE_CACHE_URLS']:
                region_urls = config['DOGPILE_CACHE_URLS']
            else:
                raise ValueError("`DOGPILE_CACHE_URLS` must be specified")

            if len(region_tuple) > 4:
                region_arguments = region_tuple[4]
                if not isinstance(region_arguments, dict):
                    raise ValueError('`regions_arguments` must be dict')
            else:
                region_arguments = config['DOGPILE_CACHE_ARGUMENTS']

            arguments = dict(url=region_urls)
            arguments.update(region_arguments)

            region = make_region().configure(
                backend=region_backend,
                expiration_time=region_timeout,
                arguments=arguments,
                wrap=wrappers_debug if app.debug else wrappers_production,
            )
            region_decorator = region.cache_on_arguments()

            self._cache_regions[region_name] = region
            self._cache_regions_decorators[region_name] = region_decorator

        if not hasattr(app, 'extensions'):
            app.extensions = {}

        app.extensions['dogpile_cache'] = self

    def get_region(self, region_name):
        """
        Method for getting already configured cache region.

        :param region_name: Name for region from config (for example, 'hour').

        :return <CacheRegion object>:

        Will raise RunTimeError if call it before `init_app` method.
        """
        if isinstance(self._cache_regions, NotInitialized):
            raise RuntimeError('working outside of application context')
        return self._cache_regions[region_name]

    def get_all_regions(self):
        """
        Method for getting all already configured cache regions.

        :return dict: keys = region_names, values = <CacheRegion objects>

        Will raise RunTimeError if call it before `init_app` method.
        """
        if isinstance(self._cache_regions, NotInitialized):
            raise RuntimeError('working outside of application context')
        return self._cache_regions

    def get_region_decorator(self, region_name):
        """
        Method for getting already configured cache region_decorator.

        :param region_name: Name for region from config (for example, 'hour').

        :return <CacheRegion object>.cache_on_arguments():

        Will raise RunTimeError if call it before `init_app` method.
        """
        if isinstance(self._cache_regions_decorators, NotInitialized):
            raise RuntimeError('working outside of application context')
        return self._cache_regions_decorators[region_name]

    def region(self, name):
        """
        CacheRegion decorator.

        :param name: Region name from config['DOGPILE_CACHE_REGIONS'].

        Example:

            @cache.region('hour')
            def cached_func(*args):
                return args

            cached_value = cached_func()
        """
        def decorator(func):
            setattr(func, self.FUNC_REGION_NAME_ATTR, name)

            @wraps(func)
            def wrapper(*args):
                if name in self._cache_regions:
                    cache_decorator = self.get_region_decorator(name)
                    return cache_decorator(func)(*args)
                else:
                    raise KeyError("You didn't specified region `%s`" % name)
            return wrapper
        return decorator

    def invalidate_region(self, region_name, hard=True):
        """
        Method for invalidation cache for all funcs decorated with particular
        @cache.region().

        :param region_name: Region name from config['DOGPILE_CACHE_REGIONS'].

        :param hard: if True, cache values will all require immediate
                     regeneration; dogpile logic won't be used.  If False, the
                     creation time of existing values will be pushed back
                     before the expiration time so that a return+regen will be
                     invoked.

        Example:

            cache.invalidate_region('hour')
        """
        region = self.get_region(region_name)
        region.invalidate(hard)

    def invalidate_all_regions(self, hard=True):
        """
        Method for invalidation cache for all funcs decorated with
        @cache.region().

        :param hard: if True, cache values will all require immediate
                     regeneration; dogpile logic won't be used.  If False, the
                     creation time of existing values will be pushed back
                     before the expiration time so that a return+regen will be
                     invoked.
        Example:

            cache.invalidate_region('hour')
        """
        for region_name in self.get_all_regions().keys():
            self.invalidate_region(region_name, hard)

    def invalidate(self, func, *args):
        """
        Method for invalidating cache for particular func.

        :param func: Function, decorated with @cache.region().

        :param *args: Decorated function arguments.

        Example:

            cache.invalidate(cached_func_without_args)
            cache.invalidate(cached_func_with_args, *args)
        """
        region_name = getattr(func, self.FUNC_REGION_NAME_ATTR)
        decorator = self.get_region_decorator(region_name)
        func = decorator(func)
        return func.invalidate(*args)

    def refresh(self, func, *args):
        """
        Method for refreshing cache for particular func.

        :param func: Function, decorated with @cache.region().

        :param *args: Decorated function arguments.

        Example:

            cache.refresh(cached_func_without_args)
            cache.refresh(cached_func_with_args, *args)

        You must understand that refreshing func do not call regenerating cache
        for func. If cache exists it does nothing. If cache does not exist it
        will compute (create) new cache value and store it to cache server.
        """
        region_name = getattr(func, self.FUNC_REGION_NAME_ATTR)
        decorator = self.get_region_decorator(region_name)
        func = decorator(func)
        return func.refresh(*args)

    def set(self, func, value, *args):
        """
        Method for setting custom cache value for particular func.

        :param func: Function, decorated with @cache.region().

        :param value: Value to store.

        :param *args: Decorated function arguments.

        Example:

            cache.set(cached_func_without_args, 'value')
            cache.set(cached_func_with_args, 'value', *args)
        """
        region_name = getattr(func, self.FUNC_REGION_NAME_ATTR)
        decorator = self.get_region_decorator(region_name)
        func = decorator(func)
        return func.set(value, *args)
